docs: Add an ADR the format of README files in repos. (#363)
openedx/open-edx-proposals: oeps/processes/oep-0055/decisions/0003-readme-specification.rst

0003 README specification
#########################

Status
******

** Accepted **

Context
*******

> `OEP-55`_ states that one of the jobs of a maintainer is te maintain an

Reviewer: typo: 
```suggestion
`OEP-55`_ states that one of the jobs of a maintainer is to maintain an
```

up-to-date and useful readme. While it laid out some possibilites about what
should go into a README, it did not go into much detail.

Decision
********

> We will define a specification that READMEs in the open edx project must

Reviewer 2: ```suggestion
We will define a specification that READMEs in the Open edX project must
```

follow.  By having a single definition, we can make it easier for people to
navigate between projects and find the most important information quickly.

Specification
*************

The readme of a well-maintained repo should contain at least the following

> sections.  Ideally in this exact order.

Reviewer 3: Assuming we didn't think of all sections and usages for future (or event current READMEs), may there should be a process noted here for:
1. Where to insert new custom headers when you feel they are warranted.
2. Whether there should be a process for capturing popular optional headers to help provide consistency.

Here is an example of where I just updated a README and wanted to provide a little warning around installation, and ad-hoc made up and inserted a new header: https://github.com/openedx/event-bus-kafka/blame/main/README.rst#L89. Would you see this as a subsection on one of your proposed headers? Other?

Also, if there are standard or recommended subsections, etc. they could be detailed here. Related, these intro sentences could mention that there is a template below that may provide more clarity for each header.

* **Title** - The name of the component.

* **Badges** - A set of `badges <https://github.com/badges/shields>`_ that
  quickly let visitors review the health of the project.

	 * The badges section should contain at minimum the following badges:

		  * **License** - Indicating the license of the repo.

		  * **Maintenance Status** - One of ``maintained``, ``deprecated``, ``unsupported``, ``experimental``.

		  * **CI Status** - The status of CI testing on the main branch.

> 		  * **Package Published Status** - If the component is a library, then a badge indicating whether publishes to the relevant package index are passing. Eg. Has the latest version been successfully published to PyPI or NPM?

Reviewer 2: ```suggestion
		  * **Package Published Status** - If the component is a library, then a badge indicating whether publishes to the relevant package index are passing. For example, has the latest version been successfully published to PyPI or NPM?
```

Reviewer 2: It could be useful to show how to make the badges. You show a green "maintained" badge in the template, but we should standardize on the look of all of the statuses.

> * **Purpose** - A paragraph or two summarizing the purpose of this component.

Reviewer 2: We should emphasize here who the audience is for the purpose.  They will often need more context than the author of the repo realizes.  "This is the Dragon service built by Aperture" is not a good enough description.  Place the repo in context of the overall project.

* **Getting Started** - A section to indicate how you can begin development on

>   the new component.

Reviewer 2: Sometimes Getting Started is about starting to use the component, sometimes it's about starting to develop.  You've clearly said "development" here. I wonder if the title of the section should clarify that, or if we might also need how to start using the component?

* **Getting Help** - A section to indicate where and how the user can get help
  with the component.  It should link to relevant docs as well as places where
  the users can get help with from other humans.

> 

Reviewer 2: Do we need a section about how to report issues?

Reviewer 2: Other sections we might need:
- Supported platforms (Aspen--Nutmeg, Python 3.x?, OS?)
- Changelog or link to one?
- Status in Open edX: core, supported, optional, etc?
- Installation: where are the docs for how to start running this component?

Reviewer 2: I'm not sure what will happen with this comment?

> * **How to Contribute** - A section to indicate how new users can contribute to

Reviewer 4: Do we need to have a status or similar around whether contributions are desired and/or will be accepted?

Author: I was considering renaming this section to `Contributing` and then we could let maintainers decide what level of contributions this component is accepting.  If we did that, I could imagine us having a status for the level of contribution accepted.  Possible values being:

* fixes only - no new features are being accepted at this point, but fixes and maintenance updates welcome.
* new-features accepted - please discuss your new ideas with the maintainers before you write code to increase the changes that features are accepted.
* security only - no changes accepted except for security fixes.

@e0d what do you think, is it worth updating this?

Reviewer 4: That makes sense to me.  Though I think that we should emphasize the following:
1. contributions should always start with an issue, before code
2. we prefer maintainers who accept contributions

  this component.  This section should at the very least contain a link to the
  general `How to Contribute <https://openedx.org/r/how-to-contribute>`_
  documentation for the Open edX Project.

* **Open edX Code of Conduct** - A section that links to the `Open edX Code of
  Conduct <https://openedx.org/code-of-conduct/>`_

* **People** - A section that tells people how they can find the current
  maintainers of this component.  Usually a link to the Backstage page for the

>   component and to the catalog-info.yaml file.

Reviewer 2: Is this section duplicative of Getting Help?

Also, do we need to link to both the Backstage page and the yaml file? Isn't one the published version of the other? I haven't seen what this will look like, does it tell how to actually contact people?

* **Reporting Security Issues** - A link to the e-mail address where security
  issues can be reported.

> Template

Reviewer 4: Would a link to a template in cookie cutters be better?  Worry this will drift compared to where the rubber meets the road.

Author: That's a good call, Once we've hammerd out the contents, I'll update the cookiecutter and link to it from here.

========

.. code-block:: rst

>    ##########

Reviewer 3: Remove overline?

   .. |Status| image:: https://img.shields.io/badge/status-maintained-31c653

   Purpose
   *******

   A few sentences that tell you what this code does and when you might want to use it.

   Getting Started
   ***************

	Developing
	==========

	How can a new developer get started with developing on this component?  Is
	there a short set of commands they could run? Is it complicated and this should
	be a link out to a published larger How-To?

   .. code:: bash

   	  git clone git@github.com:openedx/the_repo.git
   	  virtualenv venv/DoneXBlock/
   	  source venv/DoneXBlock/activate
   	  make upgrade
   	  make install
   	  make dev.run

   You can interact with the DoneXBlock in the Workbench by navigating to http://localhost:8000

	Deploying
	=========

	How can a new user, go about deploying this component? Is it just a few commands? Is there a larger how-to that should be linked here?

	For details on how to deploy this component, checkout the `deployment how-to`_

	.. _deployment how-to: https://docs.openedx.org/projects/this-project/how-tos/how-to-deploy-this-component.html

   Getting Help

>    ============

Reviewer 3: Is this meant to be a subsection of "Getting Started"? Reminder to ensure that the template is in sync with the specification above. Same question for many of the sections below. I'm guessing the heading-level is just off, but it is unclear.

   If you're having trouble, we have discussion forums at
   https://discuss.openedx.org where you can connect with others in the
   community.

   Our real-time conversations are on Slack. You can request a `Slack
   invitation`_, then join our `community Slack workspace`_.

   For anything non-trivial, the best path is to open an issue in this
   repository with as many details about the issue you are facing as you
   can provide.

   https://github.com/openedx/DoneXBlock/issues

   For more information about these options, see the `Getting Help`_ page.

   .. _Slack invitation: https://openedx.org/slack
   .. _community Slack workspace: https://openedx.slack.com/
   .. _Getting Help: https://openedx.org/getting-help

   How to Contribute
   =================

   Details about how to become a contributor to the Open edX project may
   be found in the wiki at `How to contribute`_

   .. _How to contribute: https://openedx.org/r/how-to-contribute

   The Open edX Code of Conduct

>    ----------------------------

Reviewer 3: Because this heading-level is exceptional, wondering if this is intentionally a subsection? If so, the above specification should note that more clearly.

>    All community members should familarize themselves with the `Open edX Code of Conduct`_.

Reviewer 2: "All community members are expected to follow the Open edX Code of Conduct" or something like that.

   .. _Open edX Code of Conduct: https://openedx.org/code-of-conduct/

   People
   ======

   The assigned maintainers for this component and other project details

>    may be found in `Backstage`_ or groked from inspecting catalog-info.yaml.

Reviewer 2: We aren't going to say "groked", right? :)

0003 README specification
#########################

Status
******

> ** Accepted **

Reviewer 3: Nit: I generally haven't seen formatting for the status.

> Context

Reviewer 3: Oops. @feanil - I didn't mean to approve without https://github.com/openedx/open-edx-proposals/pull/363#discussion_r936684013 being addressed. It was mistakenly marked as outdated, because a separate change came through on the same line.

Author: Ah, yes, I meant to reply on that one, I just added an update to OEP-55 and OEP-19

Reviewer 3: Thanks. I think the context paragraph should note the history of addressing READMEs in OEP-19, but I won't belabor the point. I understand it wasn't part of your immediate concern, OEP-55, but it was the initial attempt at codifying the use and purpose of a README.

Specification
*************

The readme of a well-maintained repo should contain at least the following
sections.  Ideally in this exact order with additional sections added where it
makes sense based on the maintainers judgement.

* **Title** - The name of the component.

* **Badges** - A set of `badges <https://github.com/badges/shields>`_ that
  quickly let visitors review the health of the project.

	 * The badges section should contain at minimum the following badges:

		  * **License** - Indicating the license of the repo.

		  * **Maintenance Status** - One of ``maintained``, ``deprecated``, ``unsupported``, ``experimental``.

		  * **CI Status** - The status of CI testing on the main branch.

		  * **Package Published Status** - If the component is a library, then a badge indicating whether publishes to the relevant package index are passing. For example, has the latest version been successfully published to PyPI or NPM?

* **Purpose** - A paragraph or two summarizing the purpose of this component.
  This should be oriented towards people who are new to the Open edX project.

* **Getting Started** - A section that details various aspects of getting

>   started with this component.  Some common use caseses include, getting started

Reviewer 3: ```suggestion
  started with this component.  Some common use cases include, getting started
```

  with development or how to deploy this component.

* **Getting Help** - A section to indicate where and how the user can get help
  with the component, including how they should go about reporting issues with
  the component. It should link to relevant docs as well as places where the
  users can get help from other humans.

* **Contributing** - A section to indicate how new users can contribute to
  this component.  This section should at the very least contain a link to the
  general `How to Contribute <https://openedx.org/r/how-to-contribute>`_
  documentation for the Open edX Project.

  The level of contributions that the maintainer is willing to accept should
  also be clearly described here.  Some possible values might be:

   * **fixes only** - no new features are being accepted at this point, but fixes
     and maintenance updates welcome.

>    * **new-features accepted** - please discuss your new ideas with the

Reviewer 2: ```suggestion
   * **new features accepted** - please discuss your new ideas with the
```

     :target: https://github.com/openedx/DoneXBlock/actions/workflows/ci.yml

   .. |Publish package to PyPi| image:: https://github.com/openedx/DoneXBlock/actions/workflows/pypi-release.yml/badge.svg
     :target: https://github.com/openedx/DoneXBlock/actions/workflows/pypi-release.yml

   .. TODO Link to how we make these and have examples for the other standard states.
   .. |Status| image:: https://img.shields.io/badge/status-maintained-31c653

   Purpose
   *******

   A few sentences that tell you what this code does and when you might want to use it.

   Getting Started
   ***************

   Developing
   ==========

   How can a new developer get started with developing on this component?  Is
   there a short set of commands they could run? Is it complicated and this should
   be a link out to a published larger How-To?

   .. code:: bash

        git clone git@github.com:openedx/the_repo.git
        virtualenv venv/DoneXBlock/
        source venv/DoneXBlock/activate
        make upgrade
        make install
        make dev.run

   You can interact with the DoneXBlock in the Workbench by navigating to http://localhost:8000

   Deploying
   =========

>    How can a new user, go about deploying this component? Is it just a few commands? Is there a larger how-to that should be linked here?

Reviewer 2: ```suggestion
   How can a new user go about deploying this component? Is it just a few commands? Is there a larger how-to that should be linked here?
```

   For details on how to deploy this component, checkout the `deployment how-to`_

   .. _deployment how-to: https://docs.openedx.org/projects/this-project/how-tos/how-to-deploy-this-component.html

   Getting Help
   ************

   If you're having trouble, we have discussion forums at
   https://discuss.openedx.org where you can connect with others in the
   community.

   Our real-time conversations are on Slack. You can request a `Slack
   invitation`_, then join our `community Slack workspace`_.

   For anything non-trivial, the best path is to open an issue in this
   repository with as many details about the issue you are facing as you

>    can provide.
> 
>    https://github.com/openedx/DoneXBlock/issues

Reviewer 2: ```suggestion
   can provide: https://github.com/openedx/DoneXBlock/issues
```

   For more information about these options, see the `Getting Help`_ page.

   .. _Slack invitation: https://openedx.org/slack
   .. _community Slack workspace: https://openedx.slack.com/
   .. _Getting Help: https://openedx.org/getting-help

   Contributing
   ************

   Details about how to become a contributor to the Open edX project may

>    be found in the wiki at `How to contribute`_

Reviewer 2: ```suggestion
   be found in the wiki at `How to contribute`_.
```
